### PR TITLE
storage rewrite

### PR DESCRIPTION
**Pull request by silicons**
new week new WIP pr, now containing better storage that addresses practically every complaint people have had about storage

this is now ready.

### CODE REVIEW

File: code/modules/hardsuits/_rig.dm [hunk 1]
```diff
@@ -121,6 +121,37 @@
...
+	var/storage_ui_numerical_mode = FALSE
+
+	/// storage datum path
+	var/storage_datum_path = /datum/object_system/storage
+	/// Cleared after Initialize().
+	/// List of types associated to amounts.
+	var/list/storage_starts_with
+	/// set to prevent us from spawning starts_with
+	var/storage_empty = FALSE
+
```
Comment (Zandario) [inline code/modules/hardsuits/_rig.dm:154]: With the amount of vars you need to add to something for this system, we might as well contain it inside another datum...

File: code/__DEFINES/combat/atom_damage.dm [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
+#define ATOM_DECONSTRUCT_DISASSEMBLED (1<<0)
 /// Atom received excess kinetic energy, usually through weaponry
-#define ATOM_DECONSTRUCT_DESTROYED 1
+#define ATOM_DECONSTRUCT_DESTROYED (1<<1)
 /// Atom was the victim of a corrosive or caustic chemical
-#define ATOM_DECONSTRUCT_ACID 2
+#define ATOM_DECONSTRUCT_ACID (1<<2)
 /// Atom was put out of commission by an unexpected thermal event
-#define ATOM_DECONSTRUCT_FIRE 3
+#define ATOM_DECONSTRUCT_FIRE (1<<3)
```
Comment (Zandario) [inline code/__DEFINES/combat/atom_damage.dm:13]: Really shouldn't be bitflags.

**Review by Zandario — COMMENTED** [review]
Pending HUD positioning fix

File: code/modules/hardsuits/_rig.dm (hunk 1 above)
Comment (silicons) [reply code/modules/hardsuits/_rig.dm:154]: rigsuits update deals with it, and it actually does contain it in another datum

this is a blocker for rigsuits update :^)

File: code/modules/assembly/holder.dm [hunk 3]
```diff
@@ -73,11 +73,10 @@
...
 
-/obj/item/assembly_holder/on_found(mob/finder as mob)
-	if(a_left)
-		a_left.on_found(finder)
-	if(a_right)
-		a_right.on_found(finder)
+/obj/item/assembly_holder/on_containing_storage_opening(datum/event_args/actor/actor, datum/object_system/storage/storage)
+	. = ..()
+	. |= a_left?.on_containing_storage_opening(arglist(args))
+	. |= a_right?.on_containing_storage_opening(arglist(args))
```
[the remark this answers is not in the document]
Comment (silicons) [reply code/modules/assembly/holder.dm:79]: #5677

File: code/game/click/items.dm [hunk 4]
```diff
@@ -22,18 +22,20 @@
...
 		return
 
 	// todo: refactor
-	if(resolve_attackby(target, user, params, null, clickchain_flags))
+	if(resolve_attackby(target, user, params, null, .))
 		return CLICKCHAIN_DO_NOT_PROPAGATE
 
 	// todo: signal for afterattack here
-	return clickchain_flags | afterattack(target, user, clickchain_flags, params)
+	return . | afterattack(target, user, clickchain_flags, params)
```
[the remark this answers is not in the document]
Comment (silicons) [reply code/game/click/items.dm:38]: clickchain_flags is never mutated, . is

the point of clickchain flags is to propagate through the proc and basically gather the flags for later checking, e.g.

you can mid-way through something go "we return default but also prevent attacking"
etc

**Review by Zandario — DISMISSED** [review]
Really just need to figure out that weird issue with storage ui positioning

**Review by TheLordME — APPROVED** [review]
Yolo

**Pull request merged**